src/main/java/org/elasticsearch/cluster/DiffableUtils.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster;
+
+import com.carrotsearch.hppc.cursors.ObjectCursor;
+import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
+import com.google.common.collect.ImmutableMap;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Maps.newHashMap;
+
+public final class DiffableUtils {
+    private DiffableUtils() {
+    }
+
+    /**
+     * Calculates diffs between two ImmutableOpenMaps of Diffable objects
+     */
+    public static <T extends Diffable<T>> Diff diff(ImmutableOpenMap<String, T> before, ImmutableOpenMap<String, T> after) {
+        assert after != null && before != null;
+        Map<String, Diff> diffs = newHashMap();
+        List<String> deletes = newArrayList();
+        Map<String, T> adds = newHashMap();
+        for (ObjectCursor<String> key : before.keys()) {
+            if (!after.containsKey(key.value)) {
+                deletes.add(key.value);
+            }
+        }
+        for (ObjectObjectCursor<String, T> partIter : after) {
+            T beforePart = before.get(partIter.key);
+            if (!partIter.value.equals(beforePart)) {
+                if (beforePart == null) {
```

Comment:
I think we can check the beforePart == null out of the if(!..equals) and it will make it cleaner.